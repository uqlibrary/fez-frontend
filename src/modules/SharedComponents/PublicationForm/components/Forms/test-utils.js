import { useForm } from 'react-hook-form';
import { render, WithReduxStore } from 'test-utils';
import React from 'react';
import { formValues } from '../PublicationForm';

export const setupWrapper = (Form, props) => {
    const Wrapper = () => {
        const {
            control,
            formState: { isSubmitting },
        } = useForm();

        // eslint-disable-next-line react/prop-types
        return <Form {...{ control, isSubmitting, ...props, formValues: formValues(props?.formValues || {}) }} />;
    };

    return render(
        <WithReduxStore>
            <Wrapper />
        </WithReduxStore>,
    );
};
