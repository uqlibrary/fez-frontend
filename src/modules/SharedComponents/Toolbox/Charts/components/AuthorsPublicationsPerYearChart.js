import React from 'react';
import { PropTypes } from 'prop-types';
import Chart from './Chart';

export function labelFormatter() {
    if (!this?.userOptions) {
        return '';
    }
    const name = this.userOptions.name;
    const extras = (this.userOptions.extraInfoForLegend || '').split(', ');
    return extras.join('').length ? `${name} (${extras.join(', <br />')})` : name;
}

const AuthorsPublicationsPerYearChart = ({
    yAxisTitle = 'Total publications',
    series = [],
    categories = [],
    className,
}) => {
    const options = {
        title: {
            text: null,
        },
        chart: {
            type: 'column',
            height: 320,
            backgroundColor: 'transparent',
        },
        plotOptions: {
            column: {
                stacking: 'normal',
            },
        },
        xAxis: {
            categories: categories,
            labels: {
                rotation: -45,
                y: 18,
            },
        },
        yAxis: {
            min: 0,
            title: {
                text: yAxisTitle,
            },
            stackLabels: {
                enabled: true,
            },
        },
        legend: {
            align: 'right',
            verticalAlign: 'top',
            layout: 'vertical',
            x: -30,
            y: -10,
            floating: true,
            shadow: false,
            labelFormatter,
        },
        series,
    };

    return <Chart className={`${className || ''} authors-publications-per-year-chart`} chartOptions={options} />;
};
AuthorsPublicationsPerYearChart.propTypes = {
    yAxisTitle: PropTypes.string,
    series: PropTypes.array,
    categories: PropTypes.array,
    className: PropTypes.string,
};

export default React.memo(AuthorsPublicationsPerYearChart);
