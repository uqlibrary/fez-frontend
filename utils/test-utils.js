/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';

import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';

import mediaQuery from 'css-mediaquery';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');
const mime = require('mime-types');

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/dom';
import preview from 'jest-preview';

export const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <LocalizationProvider dateAdapter={AdapterMoment}>{props.children}</LocalizationProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => {
    return {
        user: userEvent.setup(),
        ...render(ui, { wrapper: AllTheProviders, ...options }),
    };
};

export const WithRouter = ({ children, route = '/', initialEntries = [route] }) => {
    const routes = [{ path: route, element: children }];
    const router = createMemoryRouter(routes, { initialEntries: initialEntries });
    return <RouterProvider router={router} />;
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

export const assertTooltipText = async (trigger, tooltipText) => {
    expect(trigger).toBeInTheDocument();
    await userEvent.hover(trigger);
    await waitFor(() => {
        expect(screen.getByRole('tooltip', { name: tooltipText, hidden: true })).toBeVisible();
    });
};

const extensionToMimeMap = {
    tiff: 'image/tiff',
    wav: 'audio/wave',
    zip: 'application/zip',
};
const mimeType = filename => extensionToMimeMap[filename.split('.').pop()];

export const defaultDatastream = {
    dsi_embargo_date: null,
    dsi_open_access: 1,
    dsi_label: '',
    dsi_copyright: null,
    dsi_state: 'A',
    dsi_size: 27932352,
};

export const createFezDatastreamInfoArray = (datastreams, pid = null, withPreview = true) => {
    let processed = datastreams.filter(datastream => typeof datastream === 'object' || typeof datastream === 'string');
    if (withPreview) {
        processed = [];
        datastreams.forEach(datastream => {
            processed.push(datastream);
            const filename = typeof datastream === 'object' ? datastream.dsi_dsid : datastream;
            const mimetype = mimeType(filename);
            // bail in case it's not derivable
            if (
                !filename.includes('.pdf') &&
                !mimetype.includes('image') &&
                !mimetype.includes('audio') &&
                !mimetype.includes('video')
            ) {
                return;
            }
            // eslint-disable-next-line no-nested-ternary
            const derivativeMimetype = mimetype.includes('audio')
                ? 'audio/mp3'
                : mimetype.includes('video')
                ? 'video/mp4'
                : 'image/jpg';

            (derivativeMimetype.includes('image') ? ['preview_', 'thumbnail_', 'web_'] : ['']).forEach(prefix => {
                processed.push({
                    dsi_dsid: `${prefix}${filename.split('.')[0]}_t.${derivativeMimetype.split('/').pop()}`,
                    dsi_mimetype: derivativeMimetype,
                });
            });
        });
    }

    return processed.map((datastream, index) => ({
        ...defaultDatastream,
        ...{
            dsi_id: index + 1,
            dsi_pid: pid,
            dsi_order: index + 1,
            dsi_mimetype: mime.lookup(typeof datastream === 'object' ? datastream.dsi_dsid : datastream),
        },
        ...(typeof datastream === 'object'
            ? datastream
            : {
                  dsi_dsid: datastream,
              }),
    }));
};

export const withDatastreams = (sources, datastreams, callback) =>
    datastreams.map(item => {
        const source = sources.find(datastream =>
            datastream.dsi_dsid.includes(
                item.dsi_dsid
                    .replace('_t.', '.')
                    .replace('preview_', '')
                    .replace('thumbnail_', '')
                    .replace('web_', '')
                    .split('.')[0],
            ),
        );
        const isDerivative = item.dsi_dsid !== source.dsi_dsid;
        callback(item, source, isDerivative);
    });

export const getDatastreamByFilename = (filename, datastreams) =>
    datastreams.find(datastream => datastream.dsi_dsid === filename);

export const createMatchMedia = width => {
    return query => ({
        matches: mediaQuery.match(query, { width }),
        /* istanbul ignore next */
        addListener: () => {},
        /* istanbul ignore next */
        removeListener: () => {},
    });
};

export const getFilenameExtension = filename => filename.split('.').pop();
export const getFilenameBasename = filename =>
    filename.replace(new RegExp(`/\.${getFilenameExtension(filename)}$/`), '');
export const addFilesToFileUploader = files => {
    const { screen, fireEvent } = reactTestingLib;
    // create a list of Files
    const fileList = files.map(file => {
        // if file it's a string, treat it as a filename
        if (typeof file === 'string') {
            return new File([getFilenameBasename(file)], file, { type: `image/${getFilenameExtension(file)}` });
        }
        // otherwise expect it to be a object with filename and mimeType keys
        return new File([getFilenameBasename(file.filename)], file.filename, { type: file.mimeType });
    });
    // drag and drop files
    fireEvent.drop(screen.getByTestId('fez-datastream-info-input'), {
        dataTransfer: {
            files: fileList,
            types: ['Files'],
        },
    });
};
export const setFileUploaderFilesToClosedAccess = async (files, timeout = 500) => {
    const { fireEvent } = reactTestingLib;
    // set all files to closed access
    for (const file of files) {
        const index = files.indexOf(file);
        await waitFor(() => screen.getByText(new RegExp(getFilenameBasename(file))), { timeout });
        fireEvent.mouseDown(screen.getByTestId(`dsi-open-access-${index}-select`));
        fireEvent.click(screen.getByRole('option', { name: 'Closed Access' }));
    }
};

export const assertEnabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeEnabled());
export const assertDisabled = (resolver, ...items) => items.forEach(item => expect(resolver(item)).toBeDisabled());

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    withRedux,
    AllTheProviders,
    WithReduxStore,
    assertTooltipText,
    WithRouter,
    createFezDatastreamInfoArray,
    getDatastreamByFilename,
    withDatastreams,
    createMatchMedia,
    preview,
    userEvent,
    assertEnabled,
    assertDisabled,
    getFilenameExtension,
    getFilenameBasename,
    addFilesToFileUploader,
    setFileUploaderFilesToClosedAccess,
};
