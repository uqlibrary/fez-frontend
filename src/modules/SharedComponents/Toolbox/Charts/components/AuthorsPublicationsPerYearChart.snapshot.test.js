import React from 'react';
import AuthorsPublicationsPerYearChart, { labelFormatter } from './AuthorsPublicationsPerYearChart';
import { rtlRender } from 'test-utils';
// import currentAuthorStats from 'mock';

const defaultProps = {
    className: 'barChart',
    /* eslint-disable max-len */
    /* prettier-ignore */
    'series': [
        { 'name': 'Journal Article', 'data': [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 75, 1, 1, 2, 3, 0, 2, 5, 3, 1, 4, 0, 1, 3, 0, 2, 3, 6, 3, 4, 9, 4, 13, 9, 10, 3, 4, 5, 7, 0, 1, 0, 0, 0, 1, 5, 1, 1] },
        { 'name': 'Conference Paper', 'data': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 4, 0, 7, 4, 8, 3, 1, 3, 3, 4, 1, 5, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0] },
        { 'name': 'Book', 'data': [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 3, 17, 0, 0, 0, 0, 0, 1, 0, 2, 4, 1, 0, 3, 0, 0] },
        { 'name': 'Conference Proceedings', 'data': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0] },
        { 'name': 'Other', 'data': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0], 'extraInfoForLegend': 'Thesis, Working Paper, Audio Document, Generic Document, Book Chapter, Newspaper Article' },
    ],
    /* prettier-ignore */
    'categories': [111, 1011, 1111, 1121, 1212, 1213, 1222, 1223, 1231, 1233, 1234, 1901, 1987, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2022, 2131, 2222, 2323, 3424],
    /* eslint-enable max-len */
    yAxisTitle: 'Total publications',
};

function setup(testProps = {}) {
    const props = {
        ...defaultProps,
        ...testProps,
    };
    return rtlRender(<AuthorsPublicationsPerYearChart {...props} />);
}

describe('AuthorsPublicationsPerYearChart ', () => {
    it('should render empty chart component', () => {
        const { container } = setup({ series: [], categories: [], yAxisTitle: 'title' });
        expect(container).toMatchSnapshot();
    });

    it('should render component with no set class name', () => {
        const { container } = setup({
            className: '',
        });
        expect(container).toMatchSnapshot();
    });

    it('labelFormatter', () => {
        expect(labelFormatter.apply()).toEqual('');
        const scope1 = {
            userOptions: {
                name: 'test',
            },
        };
        expect(labelFormatter.apply(scope1)).toEqual('test');
        const scope2 = {
            userOptions: {
                name: 'test',
                extraInfoForLegend: 'a, b',
            },
        };
        expect(labelFormatter.apply(scope2)).toEqual('test (a, <br />b)');
    });
});
