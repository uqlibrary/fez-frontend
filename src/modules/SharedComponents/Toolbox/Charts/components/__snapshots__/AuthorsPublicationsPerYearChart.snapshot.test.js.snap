// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorsPublicationsPerYearChart  should render component with no set class name 1`] = `
<div>
  <div
    class=" authors-publications-per-year-chart"
  />
</div>
`;

exports[`AuthorsPublicationsPerYearChart  should render empty chart component 1`] = `
<div>
  <div
    class="barChart authors-publications-per-year-chart"
  />
</div>
`;

exports[`AuthorsPublicationsPerYearChart  should render with default props 1`] = `
<div>
  <div
    class=" authors-publications-per-year-chart"
  />
</div>
`;
